### PR TITLE
Dashboard/Folder permission fix session

### PR DESCRIPTION
**Pull request by Jguer**
**What this PR does / why we need it**:

Fix issues from bug bash.

- Resolve inherited permissions 
- Add inherited permission resolver for role routes

### CODE REVIEW

File: pkg/services/accesscontrol/resourcepermissions/options.go [hunk 1]
```diff
@@ -42,4 +42,6 @@ type Options struct {
 	UidSolver UidSolver
 	// InheritedScopesSolver if configured can generate additional scopes that will be used when fetching permissions for a resource
 	InheritedScopesSolver InheritedScopesSolver
+	// InheritedScopePrefixes if configured are used to create evaluators with the scopes returned by InheritedScopesSolver
+	InheritedScopePrefixes []string
 }
```
Comment (eleijonmarck) [inline pkg/services/accesscontrol/resourcepermissions/options.go:47]: this is a great set of attributes. Could we explain inheritedScopes a bit more clearer here.

Or at least I am looking for something where we can explain it a bit more in detail and even with a clear example. If someone comes into this permissions system, this gets really hairy quick. ( at least in my opionon ) and these are one of those places.

Maybe you could find a better place to place clear documentation of what inherited permissions mean for children and parents and the distinction between them

**Review by gamab — APPROVED** [review]

File: pkg/services/accesscontrol/resourcepermissions/api.go [hunk 2]
```diff
@@ -30,18 +30,45 @@ func newApi(ac accesscontrol.AccessControl, router routing.RouteRegister, manage
...
 	auth := middleware.Middleware(a.ac)
 	uidSolver := solveUID(a.service.options.UidSolver)
+	inheritanceSolver := solveInheritedScopes(a.service.options.InheritedScopesSolver)
 	disable := middleware.Disable(a.ac.IsDisabled())
 	a.router.Group(fmt.Sprintf("/api/access-control/%s", a.service.options.Resource), func(r routing.RouteRegister) {
+		actionRead := fmt.Sprintf("%s.permissions:read", a.service.options.Resource)
+		actionWrite := fmt.Sprintf("%s.permissions:write", a.service.options.Resource)
 		scope := accesscontrol.Scope(a.service.options.Resource, a.service.options.ResourceAttribute, accesscontrol.Parameter(":resourceID"))
-		actionWrite, actionRead := fmt.Sprintf("%s.permissions:write", a.service.options.Resource), fmt.Sprintf("%s.permissions:read", a.service.options.Resource)
+		readEvaluator, writeEvaluator := a.getEvaluators(actionRead, actionWrite, scope)
```
[the remark this answers is not in the document]
Comment (gamab) [reply pkg/services/accesscontrol/resourcepermissions/api.go:66]: In my option, they are different and I'd rather have them named differently to avoid the confusion.
* `evaluator` is the name given to the structure able to evaluate a user has been granted a set of permissions, it is used by the accesscontrol authorize middleware.
* here `solver` is a middleware computing something for later usage. The `uidSolver` defined here replaces the value stored in the reqContext ":resourceID" from UID to ID instead. The `inheritanceSolver` adds inherited scopes (folder scopes in this specific case) to the reqContext.

[the remark this answers is not in the document]
Comment (eleijonmarck) [reply pkg/services/accesscontrol/resourcepermissions/api.go:66]: @gamab 
if this is the case then i get that distinction!
do we use this pattern elsewhere? maybe we do actually if im not mistaken. 🤔

File: pkg/services/accesscontrol/resourcepermissions/api.go [hunk 3]
```diff
@@ -30,18 +30,45 @@ func newApi(ac accesscontrol.AccessControl, router routing.RouteRegister, manage
...
+		// Scopes in the request context parameters are to be added by solveInheritedScopes.
+		// If a user got actionRead on any of the inherited scopes, they will be granted access to the endpoint.
+		// Ex: a user inherits dashboards:read from the containing folder (folders:uid:BCeknZL7k)
+		inheritedRead := []accesscontrol.Evaluator{accesscontrol.EvalPermission(actionRead, scope)}
+		inheritedWrite := []accesscontrol.Evaluator{accesscontrol.EvalPermission(actionWrite, scope)}
+		for _, scopePrefix := range a.service.options.InheritedScopePrefixes {
+			inheritedRead = append(inheritedRead,
+				accesscontrol.EvalPermission(actionRead, accesscontrol.Parameter(scopePrefix)))
+			inheritedWrite = append(inheritedWrite,
+				accesscontrol.EvalPermission(actionWrite, accesscontrol.Parameter(scopePrefix)))
```
Comment (IevaVasiljeva) [inline pkg/services/accesscontrol/resourcepermissions/api.go:48]: question: how do these inherited scope prefixes get resolved to scopes with UIDs? I'm finding it a bit hard to follow the flow with all the middlewares.

File: pkg/services/accesscontrol/resourcepermissions/options.go (hunk 1 above)
Comment (IevaVasiljeva) [reply pkg/services/accesscontrol/resourcepermissions/options.go:47]: I agree that it's difficult to wrap your head around. I'm not sure, however, if this is the right place for more detailed documentation.

Once we have a bit more time, we will either try to simplify the access control flow or add proper documentation with diagrams to it. But for now I think we should just go ahead and fix the issues before the release. I know it's creating tech debt for future, but we need to move fairly quickly right now.

File: pkg/services/accesscontrol/resourcepermissions/api.go (hunk 3 above)
Comment (IevaVasiljeva) [reply pkg/services/accesscontrol/resourcepermissions/api.go:48]: answer: it's added to context parameters by `solveInheritedScopes` and then resolved from context parameters by `accesscontrol.Parameter`.

**Review by IevaVasiljeva — APPROVED** [review]
Looks good to me, thanks for fixing!

I think in longer term we should rethink this approach - as Eric said, here we're adding some more complexity by introducing yet another new mechanic for FGAC (passing scopes around in context). But I think it's ok as a fix for now, and we will refactor code to make it more consistent after 9.0.
